### ISSUE TITLE
[FEA] Support partitioning directories in reading / writing parquet

### ISSUE THREAD

**Issue by kkraus14** [rapidsai/cudf#2845]
Pandas / PyArrow supporting reading / writing partitioned parquet directories, example:
```
import pandas as pd

pdf = pd.DataFrame({'year': [2019, 2019, 2018, 2018], 'month': [0, 1, 0, 1], 'data': [1, 2, 3, 4]})
########################
#    year  month  data #
# 0  2019      0     1 #
# 1  2019      1     2 #
# 2  2018      0     3 #
# 3  2018      1     4 #
########################

pdf.to_parquet('test', partition_cols=['year', 'month'])
```

This produces a directory structure like the following:
```
test
├── year=2018
│   ├── month=0
│   │   └── 292a94a60e574e9e921d5e4d2e9270e8.parquet
│   └── month=1
│       └── 4f27b378a8eb462494a7c2416e1bfb9e.parquet
└── year=2019
    ├── month=0
    │   └── 5db3392f70ec42fd9cadbfe961a3ce0b.parquet
    └── month=1
        └── a68f134589ea4e1ca5fdae1b515ea99f.parquet
```

It would be great if we could handle this natively from the C++ side given a directory for both Parquet and possibly ORC files.

cc @rapidsai/cudf-java-codeowners would Spark write / ready files in a similarly partitioned way?

**Comment by jlowe** [comment]
Yes Spark would write and read that Parquet input with the understanding that year and month are partition keys.  We currently handle partitioned key values as scalars that are filled to full columns and collected with the GPU Parquet parser result to form the final parsed table.

**Comment by tgravescs** [comment]
Like Jason said, yes.  Here is a parquet doc on Spark for the partition discovery that is in open source Spark - https://spark.apache.org/docs/latest/sql-data-sources-parquet.html#partition-discovery

For writing the command in spark would just be something like:
df.write.partitionBy("year", "month", "day").parquet("/path/to/output")

**Comment by kkraus14** [comment]
Would you expect the C++ call to handle being able to be passed a directory and returning a cudf::table based on all of the parquet files in that directory? If not, would that simplify / help your codepath? I believe we're going to need to handle this from Python regardless but trying to understand others expectations to see if we're the outlier here.

**Comment by jlowe** [comment]
I agree it's definitely needed for the pandas-like use-case, but currently the Spark plugin would not leverage this feature.  The Spark plugin is always parsing data on the GPU with host buffers rather than file paths because:
* Spark loads from many types of distributed filesystems, some potentially with only Java clients
* Spark can load data that is compressed with codecs not currently supported in cuDF (e.g.: lzo, zstd)
* cuDF parsers do not handle the Hadoop-style split processing of large, individual files

The Spark driver performs data schema discovery as part of query planning, including determining which fields are partitioned.  Then when a job task loads its split of a file, it performs the following steps:
1. CPU parses the Parquet footer to determine which row groups are necessary.  Predicate push-down filters are leveraged here to avoid loading rows that are not relevant to the query.
1. CPU loads the necessary raw Parquet row group data into a host buffer
1. CPU fabricates a new Parquet footer at the end of the host buffer corresponding to what was loaded into the host buffer
1. GPU Parquet parser unpacks the Parquet rows from the host buffer "file"
1. Partitioned values, if any, become GPU scalars splatted to individual columns and collected with the GPU parse result to form the final table

Similar approaches are used for ORC and CSV input formats, so we never end up passing a directory path to the cudf parser.

**Comment by kkraus14** [comment]
Thanks for the great explanation!

**Comment by GregoryKimball** [comment]
Closed by #10000
